service_test.go

```diff
@@ -0,0 +1,322 @@
+	}{
+		// Test that adding a single subnet returns the correct subnet.
+		{
+			network: "10.4.0.0/16",
+			steps: []step{
+				{
+					add:            true,
+					mask:           24,
+					expectedSubnet: "10.4.0.0/24",
+				},
+			},
+		},
+
+		// Test that adding three subnets returns the correct subnets.
+		{
+			network: "10.4.0.0/16",
+			steps: []step{
+				{
+					add:            true,
+					mask:           24,
+					expectedSubnet: "10.4.0.0/24",
+				},
+				{
+					add:            true,
+					mask:           24,
+					expectedSubnet: "10.4.1.0/24",
+				},
+				{
+					add:            true,
+					mask:           24,
+					expectedSubnet: "10.4.2.0/24",
+				},
+			},
+		},
+
+		// Test that adding two subnets with different mask sizes,
+		// returns the correct subnets.
+		{
+			network: "10.4.0.0/16",
+			steps: []step{
+				{
+					add:            true,
+					mask:           25,
+					expectedSubnet: "10.4.0.0/25",
+				},
+				{
+					add:            true,
+					mask:           26,
+					expectedSubnet: "10.4.0.128/26",
+				},
+			},
+		},
+
+		// Test adding a network that is too large returns an error.
+		{
+			network: "10.4.0.0/16",
+			steps: []step{
+				{
+					add:                  true,
+					mask:                 15,
+					expectedErrorHandler: IsMaskTooBig,
+				},
+			},
+		},
+
+		// Test that adding a subnet, deleting it, and adding another subnet,
+		// works correctly.
+		{
+			network: "10.4.0.0/16",
+			steps: []step{
+				{
+					add:            true,
+					mask:           24,
+					expectedSubnet: "10.4.0.0/24",
+				},
+				{
+					add:            false,
+					subnetToDelete: "10.4.0.0/24",
+				},
+				{
+					add:            true,
+					mask:           24,
+					expectedSubnet: "10.4.0.0/24",
+				},
+			},
+		},
+
+		// Test that adding two subnets, deleting the first one, then
+		// adding a third larger subnet, and then a fourth of the original size,
+		// works correctly.
+		{
+			network: "10.4.0.0/16",
+			steps: []step{
+				{
+					add:            true,
+					mask:           24,
+					expectedSubnet: "10.4.0.0/24",
+				},
+				{
+					add:            true,
+					mask:           24,
+					expectedSubnet: "10.4.1.0/24",
+				},
+				{
+					add:            false,
+					subnetToDelete: "10.4.0.0/24",
+				},
+				{
+					add:            true,
+					mask:           23,
+					expectedSubnet: "10.4.2.0/23",
+				},
+				{
+					add:            true,
+					mask:           24,
+					expectedSubnet: "10.4.0.0/24",
+				},
+			},
+		},
+	}
+
+	for index, test := range tests {
+		// Parse network.
+		_, network, err := net.ParseCIDR(test.network)
+		if err != nil {
+			t.Fatalf("%v: error returned parsing network cidr: %v", index, err)
+		}
+
+		// Create a new IPAM service.
+		config := DefaultConfig()
+		config.Network = network
+
+		service, err := New(config)
+		if err != nil {
+			t.Fatalf("%v: error returned creating ipam service: %v", index, err)
+		}
+
+		for _, step := range test.steps {
```

Comment:
The error handling test runners are super gross. Don't see how to improve this, I rewrote it like, 4 times already :D